### ISSUE TITLE
D418 edge case: renaming arguments in `overload`

### ISSUE THREAD

**Issue by jamesbraza** [astral-sh/ruff#11643]
With `ruff==0.4.6`, [D418](https://docs.astral.sh/ruff/rules/overload-with-docstring/) will get thrown when there is a docstring in a `@overload`'d signature. This makes sense in the most common case where there's some type variation but no argument renames.

However, imagine the argument names themselves change too, like in the below (inspired by the first snippet in https://peps.python.org/pep-0484/#function-method-overloading):

```python
from __future__ import annotations

from typing import overload

class Foo:
    @overload
    def __getitem__(self, i: int) -> int:
        """
        Do something unique with an index.

        Args:
            i: Index.
        """

    @overload
    def __getitem__(self, s: slice) -> bytes:
        """
        Do something else with a slice.

        Args:
            s: Slice.
        """

    def __getitem__(self, *args, **kwargs):
        """Do something."""
```

Currently, Ruff will trigger `D418` on the overloads, saying functions with `@overload` shouldn't contain a docstring.

However, I think that `D418` should not be triggered, because the argument names themselves change, and this it's impossible to construct a proper docstring in final implementation of the method.

What do you think?

**Comment by charliermarsh** [comment]
That's interesting... My first reaction is that I agree with you but interested to see what others think.

**Comment by alicederyn** [comment]
VS Code shows per-overload docstrings. I think this check should probably be removed 🙏

**Comment by jonathandung** [comment]
What does `sphinx.ext.napoleon` make of this though?

**Comment by jamesbraza** [comment]
> What does `sphinx.ext.napoleon` make of this though?

It just renders (with the Alabaster theme) the final non-overload docstring. The overload docstrings, even if the final non-overload docstring is removed, are not rendered.

<img width="960" height="309" alt="screenshot of `sphinx.ext.napoleon` rendering" src="https://github.com/user-attachments/assets/308d12cf-34af-4875-baf3-c3adeb6c99b3" />

**Comment by jonathandung** [comment]
I tried it with `autoapi.extension` (Sphinx AutoAPI). When I removed the non-overloaded implementation, the only docstring for the overload taking a slice was rendered. So the last one takes precedence, regardless of whether there is an `@overload` on it.
